src/kernel/standard/getcwd.rs

```diff
@@ -1,41 +1,56 @@
-use crate::errors::Result;
-use crate::kernel::exit::SyscallExitResult;
+use std::os::unix::prelude::OsStrExt;
 
-pub fn enter() -> Result<()> {
-    Ok(())
+use libc::c_void;
+
+use crate::errors::*;
+
+use crate::filesystem::Translator;
+use crate::process::tracee::Tracee;
+use crate::register::PtraceWriter;
+use crate::register::{Original, SysArg, SysArg1, SysArg2, SysResult};
 
-    //set_sysnum(tracee, PR_void);
-    //status = 0;
+pub fn enter(tracee: &mut Tracee) -> Result<()> {
+    tracee
+        .regs
+        .cancel_syscall("Cancel getcwd and return our own value");
+    Ok(())
 }
 
-pub fn exit() -> SyscallExitResult {
-    //    char path[PATH_MAX];
-    //    size_t new_size;
-    //    size_t size;
-    //    word_t output;
-    //
-    //    size = (size_t) peek_reg(tracee, ORIGINAL, SYSARG_2);
-    //    if (size == 0) {
-    //        status = -EINVAL;
-    //        break;
-    //    }
-    //
-    //    /* Ensure cwd still exists.  */
-    //    status = translate_path(tracee, path, AT_FDCWD, ".", false);
-    //    if (status < 0)
-    //        break;
-    //
-    //    new_size = strlen(tracee->fs->cwd) + 1;
-    //    if (size < new_size) {
-    //        status = -ERANGE;
-    //        break;
-    //    }
-    //
-    //    /* Overwrite the path.  */
-    //    output = peek_reg(tracee, ORIGINAL, SYSARG_1);
-    //    status = write_data(tracee, output, tracee->fs->cwd, new_size);
-    //    if (status < 0)
-    //        break;
-    //
-    SyscallExitResult::None
+pub fn exit(tracee: &mut Tracee) -> Result<()> {
+    let buf_addr = tracee.regs.get(Original, SysArg(SysArg1));
+    let size = tracee.regs.get(Original, SysArg(SysArg2)) as usize;
+    if buf_addr == 0 || size == 0 {
+        return Err(Error::errno(Errno::EINVAL));
+    }
+
+    let fs_r = tracee.fs.borrow();
+    let guest_path = fs_r.get_cwd();
+    // we need to ensure cwd still exists
+    tracee
+        .fs
+        .borrow()
+        .translate_absolute_path(&guest_path, true)?
+        .metadata()?;
+
+    let bytes = guest_path.as_os_str().as_bytes();
+    let real_size = bytes.len() + 1;
+    if real_size > size {
+        return Err(Error::errno(Errno::ERANGE));
+    }
+
+    tracee
+        .regs
+        .write_data(buf_addr as *mut c_void, bytes, true)?;
+
+    // Unlike the description in the getcwd() man page, in the Linux kernel, the
+    // getcwd() system call returns the length of the buffer filled.
+    // https://elixir.bootlin.com/linux/v5.10.43/source/fs/d_path.c#L412
```

Comment:
Is this specific to v5.10.43 of the Linux kernel? Which version did this change?

Reply:
This behavior seems to have been present in the [source code](https://elixir.bootlin.com/linux/2.1.92/source/fs/dcache.c#L762) from the beginning, when [`getcwd()`](https://man7.org/linux/man-pages/man2/getcwd.2.html) was first time available in Linux 2.1.92.